[Dynamo] Support torch dynamo for PrivateUse1 backend

fmo-mt opened the pull request:
This PR enable that torch.dynamo to be used in 3rd party backends.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela @NmomoN @mengpenghui @fwenguang @cdzhan @1274085042 @PHLens @albanD

fmo-mt committed:
Support torch dynamo for PrivateUse1 backend

fmo-mt commented:
@pytorchbot label "module: dynamo"

fmo-mt commented:
@anijain2305 Hi, can you help to take a look at this, please?

changliang-wang commented:
@fmo-mt Hello, I saw you on the open source project TexasSolver and I have some questions. Can I contact you to communicate with you？

fmo-mt commented:
> @fmo-mt Hello, I saw you on the open source project TexasSolver and I have some questions. Can I contact you to communicate with you？

I don't get it, I'm not the contributer of that repo and it's pytorch here.

fmo-mt commented:
@anijain2305 Hi, can you help to take a look at this, please?

anijain2305 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -349,7 +351,7 @@ def call_function(
             torch.cuda.amp.autocast,
             torch.cpu.amp.autocast,
-        ):
+        ) + tuple(([custom_backend_mod.amp.autocast,] if custom_backend_mod else [])):
```
Consider adding a new function in _dynamo/utils.py - `is_amp_autocast` that does the custom_backend calculation once (perhaps use lru_cache).

anijain2305 reviewed (commented):
I have some simple refactor suggestions, but overall Dynamo portions look good.

I am also unfamiliar with privateuse backend. Relying on @albanD @soulitzer for c10 related changes, specially TensorImpl.h.

fmo-mt commented on torch/_dynamo/variables/ctx_manager.py:
```diff
@@ -874,11 +874,15 @@ def fn_name(self):
 class AutocastModeVariable(ContextWrappingVariable):
     @staticmethod
     def create(func, args, kwargs):
+        if custom_backend_mod := getattr(torch, torch._C._get_privateuse1_backend_name(), None):
```
Yes we can do that.

fmo-mt commented on torch/_inductor/codecache.py:
```diff
@@ -200,7 +200,7 @@ def get_system() -> dict[str, Any]:
                 system["device"]["name"] = device_properties.gcnArchName
                 system["version"]["hip"] = torch.version.hip
-        except (AssertionError, RuntimeError):
+        except (AttributeError, AssertionError, RuntimeError):
```
Removed.

fmo-mt committed:
Update

fmo-mt commented:
I modified this PR a liittle with your suggestions, and for `TensorImpl.h` I removed it from this PR and we can discuss it in https://github.com/pytorch/pytorch/pull/155215

anijain2305 commented:
Thanks for the changes.

fmo-mt committed:
fix lint error, avoiding circular import

fmo-mt commented:
@anijain2305 Hi, I fixed some lint errors and tried to fix the circular import in `torch/_prims/rng_prims.py`, you can see there're some duplicated codes in it, I don't quite sure that it's the best way of implementing whilst I just tried to avoiding lots of changes, please help to take a look, thanks.

fmo-mt committed:
fix lint error

fmo-mt commented:
@anijain2305 Hi, I fixed some lint errors(again...) and CI needs to be triggered (again, too...)

fmo-mt commented:
@pytorchbot merge

anijain2305 reviewed (changes requested):
There are some test failures. Can you take a look?

fmo-mt commented:
> There are some test failures. Can you take a look?

I tested some failure cases for MacOS and ROCm, but I cannot reproduce them, for example we have some mps tests failed, like here: https://github.com/pytorch/pytorch/actions/runs/17258181160/job/48974473646#step:16:3450, https://github.com/pytorch/pytorch/actions/runs/17258181160/job/48974473662#step:16:875 and https://github.com/pytorch/pytorch/actions/runs/17258181160/job/48974473662#step:16:875, I tested on my MacBook and a ROCm server, they run just fine. 

But on a A100 platform some tests is detected, I'm trying ot figure out what cause these failures and fix them asap.
